File: bending-api/src/main/java/pw/dotdash/bending/api/event/SetCooldownEvent.java
package pw.dotdash.bending.api.event;

import com.google.common.base.Optional;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Cancellable;
import org.spongepowered.api.event.cause.Cause;
import org.spongepowered.api.event.entity.living.humanoid.player.TargetPlayerEvent;
import org.spongepowered.api.event.impl.AbstractEvent;
import pw.dotdash.bending.api.ability.AbilityType;

import javax.annotation.Nullable;

import java.util.OptionalLong;

import static com.google.common.base.Preconditions.checkNotNull;

public class SetCooldownEvent extends AbstractEvent implements TargetPlayerEvent, Cancellable {

    private final Player targetEntity;
    private final AbilityType abilityType;
    @Nullable
    private final Long cooldownMilli;
    private final Cause cause;

    public SetCooldownEvent(Player targetEntity, AbilityType abilityType, @Nullable Long cooldownMilli, Cause cause) {
        this.targetEntity = checkNotNull(targetEntity, "targetEntity");
        this.abilityType = checkNotNull(abilityType, "abilityType");
        this.cooldownMilli = cooldownMilli;
        this.cause = checkNotNull(cause, "cause");
    }

    private boolean cancelled = false;

    @Override
    public Player getTargetEntity() {
        return this.targetEntity;
    }

    public AbilityType getAbilityType() {
        return this.abilityType;
    }

    public OptionalLong getCooldownMilli() {
        return this.cooldownMilli == null ? OptionalLong.empty() : OptionalLong.of(this.cooldownMilli);
    }

    @Override
    public Cause getCause() {
        return this.cause;
    }

    @Override
    public boolean isCancelled() {
        return this.cancelled;
    }

    @Override
    public void setCancelled(boolean cancel) {
        this.cancelled = cancel;
    }
}

File: src/main/kotlin/arvenwood/bending/api/event/SetCooldownEvent.kt
package arvenwood.bending.api.event

import arvenwood.bending.api.ability.AbilityType
import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.event.Cancellable
import org.spongepowered.api.event.cause.Cause
import org.spongepowered.api.event.entity.living.humanoid.player.TargetPlayerEvent
import org.spongepowered.api.event.impl.AbstractEvent
import java.time.Instant

data class SetCooldownEvent(
    private val targetEntity: Player,
    val type: AbilityType<*>,
    val cooldownMilli: Long?,
    private val cause: Cause
) : AbstractEvent(), TargetPlayerEvent, Cancellable {

    val cooldown: Instant? get() = this.cooldownMilli?.let(Instant::ofEpochMilli)

    private var cancelled: Boolean = false

    override fun isCancelled(): Boolean = this.cancelled

    override fun setCancelled(cancel: Boolean) {
        this.cancelled = cancel
    }

    override fun getTargetEntity(): Player = this.targetEntity

    override fun getCause(): Cause = this.cause
}